src/Compilers/CSharp/Portable/Utilities/InterceptableLocation.cs

```diff
@@ -0,0 +1,183 @@
+        }
+
+        // format:
+        // - 16 bytes of target file content hash (xxHash128)
+        // - int32 position (little endian)
+        // - utf-8 display filename
+        const int hashIndex = 0;
+        const int hashSize = 16;
+        const int positionIndex = hashIndex + hashSize;
+        const int positionSize = sizeof(int);
+        const int displayNameIndex = positionIndex + positionSize;
+        const int minLength = displayNameIndex;
+
+        if (bytes.Length < minLength)
+        {
+            diagnostics.Add(ErrorCode.ERR_InterceptsLocationDataInvalidFormat, diagnosticLocation);
+            return null;
+        }
+
+        var hash = bytes.AsMemory(start: hashIndex, length: hashSize);
+        var position = BinaryPrimitives.ReadInt32LittleEndian(bytes.AsSpan(start: positionIndex));
+
+        // https://learn.microsoft.com/en-us/dotnet/api/system.text.encoding.utf8?view=net-8.0#remarks states:
+        // > `Encoding.UTF8` returns a UTF8Encoding object that uses replacement fallback to replace each string that it can't encode and each byte that it can't decode with a question mark ("?") character.
+        //
+        // If these assertions are violated, it means the encoder may throw for invalid UTF-8 sequences, which we don't expect.
+        // In this case we would either need to adjust the code to start handling ArgumentException from GetString, or create a static UTF8 decoder with non-throwing behavior.
+        Debug.Assert(Encoding.UTF8.DecoderFallback is DecoderReplacementFallback);
+        Debug.Assert(Encoding.UTF8.IsReadOnly);
+        string displayFileName = Encoding.UTF8.GetString(bytes, index: displayNameIndex, count: bytes.Length - displayNameIndex);
+
+        return (hash, position, displayFileName);
+    }
+
+    // Note: the goal of implementing equality here is so that incremental state tables etc. can detect and use it.
+    // This encoding which uses the checksum of the referenced file may not be stable across incremental runs in practice, but it seems correct in principle to implement equality here anyway.
+    public override bool Equals(object? obj)
+    {
+        if ((object)this == obj)
+            return true;
+
+        return obj is InterceptableLocation1 other
+            && _checksum.SequenceEqual(other._checksum)
+            && _path == other._path
+            && _position == other._position
+            && _lineNumberOneIndexed == other._lineNumberOneIndexed
+            && _characterNumberOneIndexed == other._characterNumberOneIndexed;
+    }
+
+    public override int GetHashCode()
+    {
+        return Hash.Combine(
+           BinaryPrimitives.ReadInt32LittleEndian(_checksum.AsSpan()),
+           Hash.Combine(
+               _path.GetHashCode(),
+               Hash.Combine(
+                   _position,
+                   Hash.Combine(
+                       _lineNumberOneIndexed,
+                       _characterNumberOneIndexed))));
```

Comment:
Including line and character feels redundant since we're including position.

Reply:
I'll boil this down to first 4 of checksum plus position as it does seem to be enough to be "characteristic" of the location.